### ISSUE TITLE
Introduce annotations for configuring an extension

### ISSUE THREAD

**Issue by mojavelinux** [asciidoctor/asciidoctorj#196]
Introduce annotations for configuring an extension. Currently, extension configuration must be passed to the super class in the extension constructor. However, this information is mostly static and best suited for annotations in the Java world.

Here's what is currently necessary (shown in Groovy):

```groovy
class YellBlock extends BlockProcessor {
    YellBlock(String name, Map<String, Object> config) {
        super(name, [contexts: [':paragraph'], content_model: ':simple']);
    }
    ...
}
```

Here's an example of how this might look (shown in Java):

```java
@Context(":paragraph")
@ContentModel(SIMPLE)
public class YellBlock extends BlockProcessor {
  ...
}
```

There should be annotations for each of the DSL methods in the Ruby APIs.

**Comment by robertpanzer** [comment]
I looked through extensions.rb and I think that the Asciidoctor Ruby base classes currently support the config options below.
(I already translated them to dedicated annotations as in the initial approach)

* DocInfoProcessor has 
  * `@Location(String)`
    Values are `:header` and `:footer`.
    Would it make sense to define an enum for that or are other values conceivable?
* BlockProcessor & MacroProcessor & BlockMacro have
  * `@Name(String)`
  * `@Contexts(String[])`
  * `@ContentModel(String)`
  * `@PositionalAttributes(???)`
    * I could not find any values in the Asciidoctor sources...
  * `@DefaultAttributes(@DefaultAttribute[])`
    * `@DefaultAttribute(String, String)` (Does it work out to simply have Strings mapped to Strings here? Or do we need to support list etc as well?)

* InlineMacro has the same as MacroProcessor plus
  * `@UsingFormat(String)`
    * Values are `:short` and anything else (`:long`?).
       Would it make sense to define an enum for that or are other values conceivable?
  * `@Regexp(String)`

Do you think it is reasonable to create these annotations?
If so it should not be too hard to evaluate the annotations when someone registers a class at the JavaExtensionRegistry.

**Comment by mojavelinux** [comment]
> `@Location(String)` Values are :header and :footer. Would it make sense to define an enum for that or are other values conceivable?

Not currently. Though, we have been considering an additional placement at the top of the content (not the head). Technically, a converter can request any placement key it wants when resolving the docinfo. However, for now probably best to start with an enum.

> `@PositionalAttributes(???)` I could not find any values in the Asciidoctor sources...

These are attribute names, so they must be string values.

> `@DefaultAttribute(String, String)` (Does it work out to simply have Strings mapped to Strings here? Or do we need to support list etc as well?)

Simple strings as attributes can only be strings.

> `@UsingFormat(String)`

This should simply be `@Format`. Though, it really should have three values (even though this isn't implemented in Asciidoctor). It should be `:long`, `:short` or `:custom` (or `:regexp`). If `custom` (or `:regexp`, then a `@Regexp` annotation is required.

> Do you think it is reasonable to create these annotations?

**Comment by robertpanzer** [comment]
> Am 22.05.2015 um 22:05 schrieb Dan Allen <notifications@github.com>:
> 
> This should simply be @Format. Though, it really should have three values (even though this isn't implemented in Asciidoctor). It should be :long, :short or :custom (or :regexp). If custom (or :regexp, then a @Regexp annotation is required.

I think when having a @Regexp there is no need for a @Format("custom"). 
Maybe when registering an extension with @Format and @Regexp we could throw an exception that the extension is not configured correctly. 

Or maybe only @Format("long"), @Format("short") or @Format(regexp=".*")?

**Comment by mojavelinux** [comment]
> Or maybe only `@Format("long")`, `@Format("short")` or `@Format(regexp=".*")`?

Now you've got it! The default value then should be "custom" (so you don't have to specify it when you set regexp).

**Comment by robertpanzer** [comment]
One point I'd like to discuss is the following:
BlockProcessors and MacroProcessors are at the moment required to have a constructor that takes the block name or macro name.
Do we really need that?
I'd prefer to only have a no-arg constructor.
We could probably also check if the processor has a no-arg constructor or the old one and call the correct one.
Having the name in the constructor only makes sense if the processor is registered without the annotation. That would mean that it is a common case to register the same extension class with changing names.

**Comment by mojavelinux** [comment]
@robertpanzer That's precisely the idea. The argument in the constructor is optional for the case when you want to reuse the processor in another context. In other words, whatever configuration you pass to the constructor overrides the annotations, which serve as the default / fallback configuration.

**Comment by robertpanzer** [comment]
> In other words, whatever configuration you pass to the constructor overrides the annotations, which serve as the default / fallback configuration.

Yes, that's really important! 
My question was going more into the direction if it is required that there is a constructor that *receives* the block or macro name?
It can certainly invoke the super constructor passing any configuration, but it is really required to receive the name?

Maybe we could extend the logic to invoke a "specialized" constructor with the macro/block name if available and fall back to a no arg constructor?

I believe that a no arg constructor will cover 90% of the cases and that it makes it simpler getting started with writing own extensions.

**Comment by mojavelinux** [comment]
Yes, I agree. That's how I see it too. We really want the extension classes to be a simple as they possibly can be.

**Comment by robertpanzer** [comment]
Fixed with #341
